src/utils/reductions.hpp

```diff
@@ -59,20 +59,26 @@ struct ReductionBase {
   T val;
 #ifdef MPI_PARALLEL
   MPI_Request req;
-  MPI_Comm comm;
+  std::shared_ptr<MPI_Comm> comm;
 #endif
   bool active = false;
   ReductionBase() {
 #ifdef MPI_PARALLEL
-    MPI_Comm_dup(MPI_COMM_WORLD, &comm);
+    // Store the communicator in a shared_ptr, so that
+    // MPI_Comm_free is called, but only when the last
+    // copy of this ReductionBase is destroyed.
+    // TODO should Comm_free be MPI_CHECKed?
+    MPI_Comm *comm_tmp;
+    PARTHENON_MPI_CHECK(MPI_Comm_dup(MPI_COMM_WORLD, comm_tmp));
+    comm = std::shared_ptr<MPI_Comm>(comm_tmp, MPI_Comm_free);
```

Comment:
Looks like this causes a runtime error, probably because `MPI_Comm_dup` doesn't allocate. I suggest the following:

```suggestion
    comm = std::shared_ptr<MPI_Comm>(new MPI_Comm, MPI_Comm_free);
    PARTHENON_MPI_CHECK(MPI_Comm_dup(MPI_COMM_WORLD, comm.data()));
```

Reply:
With this  setup we could also put the initialization of `comm` in the var definition, e.g.,
```C++
std::shared_ptr<MPI_Comm> comm(new MPI_Comm, MPI_Comm_free);
```